code/headers/hardware_usart.hpp

```diff
@@ -2,76 +2,184 @@
 
 #include <hwlib.hpp>
 #include <queue.hpp>
-#include <uart_ports.hpp>
 #include <usart_connection.hpp>
 
-namespace r2d2 {
+namespace r2d2::usart {
 
-    class hardware_usart_c : public usart_connection_c{
+    enum class usart_ports {
+        uart1 = 0,
+        uart2,
+        uart3,
+        // only for determining how many uart ports there are
+        UART_SIZE
+    };
+
+    enum class peripheral {
+        peripheral_a,
+        peripheral_b
+    };
+
+    struct hw_usart_s {
+        Usart *usart;
+        uint32_t rx_mask;
+        uint32_t tx_mask;
+        Pio *pio;
+        peripheral periph;
+        uint16_t id;
+    };
+
+    const static hw_usart_s usart[uint8_t(usart_ports::UART_SIZE)] = {
+            {USART0, PIO_PA10A_RXD0, PIO_PA11A_TXD0, PIOA, peripheral::peripheral_a, ID_USART0},
+            {USART1, PIO_PA12A_RXD1, PIO_PA13A_TXD1, PIOA, peripheral::peripheral_a, ID_USART1},
+            {USART3, PIO_PD5B_RXD3, PIO_PD4B_TXD3, PIOD, peripheral::peripheral_b, ID_USART3}
+    };
+
+    void set_peripheral(Pio *pio, uint32_t mask, peripheral p) {
+        uint32_t t = pio->PIO_ABSR;
+
+        if (p == peripheral::peripheral_a) {
+            pio->PIO_ABSR &= (~mask & t);
+        } else {
+            pio->PIO_ABSR = (mask | t);
+        }
+
+        // remove pin from pio controller
+        pio->PIO_PDR = mask;
+    };
+
+    template <size_t BufferLength = 250>
+    class hardware_usart_c : public usart_connection_c {
     private:
         Usart *hardware_usart = nullptr;
         unsigned int baudrate;
-        bool usart_initialized = true;
-        uart_ports_c usart_port;
-        queue_c<uint8_t, 250> input_buffer;
+        queue_c<uint8_t, BufferLength> input_buffer;
 
         /// @brief check if the transmitter is ready to send
         /// @return true if ready to send, false if not ready to send
-        bool transmit_ready();
+        bool transmit_ready() {
+            return (hardware_usart->US_CSR & 2);
+        }
 
         /// @brief send a byte via usart
         /// @param b: byte to send
-        void send_byte(const uint8_t &b);
+        void send_byte(const uint8_t &b) {
+            while (!transmit_ready()) {
```

Comment:
shouldn't this be a Timeout instead of a blocking call, so it can be handled appropriately?

Reply:
Good point will fix later issue has been made #20